awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx

```diff
@@ -1,94 +1,688 @@
+  const [addingLink, setAddingLink] = useState(false);
   const [contentError, setContentError] = useState(null);
   const [isLoading, setIsLoading] = useState(true);
-  const [graphLinks, setGraphLinks] = useState([]);
-  // We'll also need to store the original set of nodes...
-  const [graphNodes, setGraphNodes] = useState([]);
+  const [linkToDelete, setLinkToDelete] = useState(null);
+  const [linkToEdit, setLinkToEdit] = useState(null);
+  const [links, setLinks] = useState([]);
+  const [nextNodeId, setNextNodeId] = useState(0);
+  const [nodePositions, setNodePositions] = useState(null);
+  const [nodeToDelete, setNodeToDelete] = useState(null);
+  const [nodeToEdit, setNodeToEdit] = useState(null);
+  const [nodeToView, setNodeToView] = useState(null);
+  const [nodes, setNodes] = useState([]);
+  const [showDeleteAllNodesModal, setShowDeleteAllNodesModal] = useState(false);
+  const [showKey, setShowKey] = useState(false);
+  const [showTools, setShowTools] = useState(false);
+  const [showUnsavedChangesModal, setShowUnsavedChangesModal] = useState(false);
+  const [unsavedChanges, setUnsavedChanges] = useState(false);
+
+  const startAddNode = (sourceNodeId, targetNodeId = null) => {
+    setAddNodeSource(sourceNodeId);
+    setAddNodeTarget(targetNodeId);
+  };
+
+  const finishAddingNode = newNode => {
+    const newNodes = [...nodes];
+    const newLinks = [...links];
+    newNodes.push({
+      id: nextNodeId,
+      type: 'node',
+      unifiedJobTemplate: newNode.nodeResource,
+    });
+
+    // Ensures that root nodes appear to always run
+    // after "START"
+    if (addNodeSource === 1) {
+      newNode.linkType = 'always';
+    }
+
+    newLinks.push({
+      source: { id: addNodeSource },
+      target: { id: nextNodeId },
+      linkType: newNode.linkType,
+      type: 'link',
+    });
+    if (addNodeTarget) {
+      newLinks.forEach(linkToCompare => {
+        if (
+          linkToCompare.source.id === addNodeSource &&
+          linkToCompare.target.id === addNodeTarget
+        ) {
+          linkToCompare.source = { id: nextNodeId };
+        }
+      });
+    }
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setAddNodeSource(null);
+    setAddNodeTarget(null);
+    setNextNodeId(nextNodeId + 1);
+    setNodes(newNodes);
+    setLinks(newLinks);
+  };
+
+  const startEditNode = node => {
+    setNodeToEdit(node);
+  };
+
+  const finishEditingNode = editedNode => {
+    const newNodes = [...nodes];
+    const matchingNode = newNodes.find(node => node.id === nodeToEdit.id);
+    matchingNode.unifiedJobTemplate = editedNode.nodeResource;
+    matchingNode.isEdited = true;
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setNodeToEdit(null);
+    setNodes(newNodes);
+  };
+
+  const cancelNodeForm = () => {
+    setAddNodeSource(null);
+    setAddNodeTarget(null);
+    setNodeToEdit(null);
+  };
+
+  const deleteNode = () => {
```

Comment:
This function is pretty long. I'd consider moving chunks of it into helper functions to it's easier to mentally track everything that's going on... probably lines 165–183 into one function ("Remove any links that reference this node") and 185–208 ("Add the new links")